compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs

```diff
@@ -93,13 +95,26 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
             previous call to `try_evaluate_added_goals!`"
         );
 
-        let certainty = certainty.unify_with(goals_certainty);
-
-        let var_values = self.var_values;
-        let external_constraints = self.compute_external_query_constraints()?;
-
+        // When normalizing, we've replaced the expected term with an unconstrained
+        // inference variable. This means that we dropped information which could
+        // have been important. We handle this by instead returning the nested goals
+        // to the caller, where they are then handled.
+        //
+        // As we return all ambiguous nested goals, we can ignore the certainty returned
+        // by `try_evaluate_added_goals()`.
+        let (certainty, normalization_nested_goals) = if self.is_normalizes_to_goal {
+            let NestedGoals { normalizes_to_goals, goals } = std::mem::take(&mut self.nested_goals);
+            assert!(normalizes_to_goals.is_empty());
```

Comment:
Can we assert that if `goal_certainty` is ambiguous, `goals.len() > 0`? Seems like a good way to make sure that "we can ignore the certainty returned by `try_evaluate_added_goals()`" remains sound.